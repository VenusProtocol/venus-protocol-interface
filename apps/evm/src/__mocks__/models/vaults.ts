import BigNumber from 'bignumber.js';

import { Vault } from 'types';

import { vai, xvs } from './tokens';

export const vaults: Vault[] = [
  {
    rewardToken: xvs,
    stakedToken: vai,
    isPaused: false,
    lockingPeriodMs: 300000,
    dailyEmissionMantissa: new BigNumber('144000000000000000000'),
    totalStakedMantissa: new BigNumber('415000000000000000000'),
    stakingAprPercentage: 12665.060240963856,
  },
  {
    rewardToken: xvs,
    stakedToken: xvs,
    isPaused: false,
    lockingPeriodMs: 300000,
    dailyEmissionMantissa: new BigNumber('144000000000000000000'),
    totalStakedMantissa: new BigNumber('400000000000000000000000000'),
    stakingAprPercentage: 12.92281835063781,
    userStakedMantissa: new BigNumber('233000000000000000000'),
  },
];
