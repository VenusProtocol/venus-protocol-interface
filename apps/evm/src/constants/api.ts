import type { Network } from 'types';

export const apiUrls: Record<Network, string> = {
  testnet: 'https://testnetapi.venus.io',
  mainnet: 'https://api.venus.io',
  'mainnet-preview': 'https://api-preview.venus.io',
};
