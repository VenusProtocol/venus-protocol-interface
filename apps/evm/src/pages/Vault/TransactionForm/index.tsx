import BigNumber from 'bignumber.js';
import { useCallback, useMemo } from 'react';

import {
  ApproveTokenSteps,
  ApproveTokenStepsProps,
  FormikSubmitButton,
  FormikTokenTextField,
  LabeledInlineContent,
  NoticeWarning,
  SpendingLimit,
} from 'components';
import { AmountForm } from 'containers/AmountForm';
import useConvertMantissaToReadableTokenString from 'hooks/useConvertMantissaToReadableTokenString';
import useTokenApproval from 'hooks/useTokenApproval';
import { displayMutationError } from 'packages/errors';
import { useTranslation } from 'packages/translations';
import { useAccountAddress } from 'packages/wallet';
import { Token } from 'types';
import { cn, convertMantissaToTokens, convertTokensToMantissa } from 'utilities';

import TEST_IDS from './testIds';

export interface TransactionFormUiProps {
  token: Token;
  tokenNeedsToBeApproved?: boolean;
  submitButtonLabel: string;
  submitButtonDisabledLabel: string;
  onSubmit: (amountMantissa: BigNumber) => Promise<unknown>;
  isSubmitting: boolean;
  availableTokensMantissa: BigNumber;
  availableTokensLabel: string;
  isTokenApproved: ApproveTokenStepsProps['isTokenApproved'];
  approveToken: ApproveTokenStepsProps['approveToken'];
  isApproveTokenLoading: ApproveTokenStepsProps['isApproveTokenLoading'];
  isWalletSpendingLimitLoading: ApproveTokenStepsProps['isWalletSpendingLimitLoading'];
  revokeWalletSpendingLimit: () => Promise<unknown>;
  isRevokeWalletSpendingLimitLoading: boolean;
  warning?: {
    amountTokens: BigNumber;
    message: string;
    submitButtonLabel: string;
  };
  walletSpendingLimitTokens?: BigNumber;
  lockingPeriodMs?: number;
}

export const TransactionFormUi: React.FC<TransactionFormUiProps> = ({
  token,
  tokenNeedsToBeApproved = false,
  availableTokensMantissa,
  availableTokensLabel,
  submitButtonLabel,
  submitButtonDisabledLabel,
  onSubmit,
  isSubmitting,
  isTokenApproved,
  approveToken,
  isApproveTokenLoading,
  isWalletSpendingLimitLoading,
  walletSpendingLimitTokens,
  revokeWalletSpendingLimit,
  isRevokeWalletSpendingLimitLoading,
  lockingPeriodMs,
  warning,
}) => {
  const { t } = useTranslation();

  const availableTokens = useMemo(
    () =>
      convertMantissaToTokens({
        value: availableTokensMantissa,
        token,
      }),
    [availableTokensMantissa, token],
  );

  const limitTokens = useMemo(() => {
    if (isTokenApproved && walletSpendingLimitTokens) {
      return BigNumber.minimum(availableTokens, walletSpendingLimitTokens);
    }

    return availableTokens;
  }, [availableTokens, isTokenApproved, walletSpendingLimitTokens]);

  const readableAvailableTokens = useConvertMantissaToReadableTokenString({
    value: availableTokensMantissa,
    token,
  });

  const readableLockingPeriod = useMemo(() => {
    if (!lockingPeriodMs) {
      return undefined;
    }

    const now = new Date();
    const unlockingDate = new Date(now.getTime() + lockingPeriodMs);

    return t('vault.transactionForm.lockingPeriod.duration', { date: unlockingDate });
  }, [lockingPeriodMs, t]);

  const shouldDisplayWarning = useCallback(
    (amountTokens: string) =>
      !!amountTokens && warning?.amountTokens.isLessThanOrEqualTo(amountTokens),
    [warning],
  );

  const handleSubmit = async (amountTokens: string) => {
    const amountMantissa = convertTokensToMantissa({
      value: new BigNumber(amountTokens),
      token,
    });

    try {
      await onSubmit(amountMantissa);
    } catch (error) {
      displayMutationError({ error });
    }
  };

  return (
    <AmountForm onSubmit={handleSubmit} maxAmount={limitTokens.toFixed()}>
      {({ dirty, isValid, setFieldValue, values }) => (
        <div className="space-y-6">
          <div>
            <FormikTokenTextField
              name="amount"
              token={token}
              disabled={isSubmitting}
              rightMaxButton={{
                label: t('vault.transactionForm.rightMaxButtonLabel'),
                onClick: () => setFieldValue('amount', limitTokens.toFixed()),
              }}
              max={limitTokens.toFixed()}
              data-testid={TEST_IDS.tokenTextField}
            />

            {warning && shouldDisplayWarning(values.amount) && (
              <NoticeWarning
                description={warning.message}
                data-testid={TEST_IDS.noticeWarning}
                className="mt-3"
              />
            )}
          </div>

          <div className="space-y-3">
            <LabeledInlineContent
              data-testid={TEST_IDS.availableTokens}
              iconSrc={token}
              label={availableTokensLabel}
            >
              {readableAvailableTokens}
            </LabeledInlineContent>

            {tokenNeedsToBeApproved && (
              <SpendingLimit
                token={token}
                walletBalanceTokens={availableTokens}
                walletSpendingLimitTokens={walletSpendingLimitTokens}
                onRevoke={revokeWalletSpendingLimit}
                isRevokeLoading={isRevokeWalletSpendingLimitLoading}
                data-testid={TEST_IDS.spendingLimit}
              />
            )}

            {readableLockingPeriod && (
              <LabeledInlineContent
                data-testid={TEST_IDS.lockingPeriod}
                label={t('vault.transactionForm.lockingPeriod.label')}
              >
                {readableLockingPeriod}
              </LabeledInlineContent>
            )}
          </div>

          {tokenNeedsToBeApproved ? (
            <ApproveTokenSteps
              token={token}
              hideTokenEnablingStep={!isValid || !dirty}
              isTokenApproved={isTokenApproved}
              approveToken={approveToken}
              isApproveTokenLoading={isApproveTokenLoading}
              isWalletSpendingLimitLoading={isWalletSpendingLimitLoading}
            >
              <FormikSubmitButton
                loading={isSubmitting}
                disabled={
                  !isValid ||
                  !dirty ||
                  isSubmitting ||
                  !isTokenApproved ||
                  isWalletSpendingLimitLoading ||
                  isRevokeWalletSpendingLimitLoading
                }
                className="w-full"
                enabledLabel={submitButtonLabel}
                disabledLabel={submitButtonDisabledLabel}
              />
            </ApproveTokenSteps>
          ) : (
            <FormikSubmitButton
              loading={isSubmitting}
              disabled={!isValid || !dirty || isSubmitting}
              enabledLabel={
                warning && shouldDisplayWarning(values.amount)
                  ? warning.submitButtonLabel
                  : submitButtonLabel
              }
              className={cn('w-full', shouldDisplayWarning(values.amount) && 'border-red bg-red')}
              disabledLabel={submitButtonDisabledLabel}
            />
          )}
        </div>
      )}
    </AmountForm>
  );
};

export interface TransactionFormProps
  extends Omit<
    TransactionFormUiProps,
    | 'isTokenApproved'
    | 'approveToken'
    | 'isApproveTokenLoading'
    | 'isWalletSpendingLimitLoading'
    | 'isWalletSpendingLimitLoading'
    | 'revokeWalletSpendingLimit'
    | 'isRevokeWalletSpendingLimitLoading'
  > {
  poolIndex?: number;
  spenderAddress?: string;
}

const TransactionForm: React.FC<TransactionFormProps> = ({
  token,
  tokenNeedsToBeApproved = false,
  spenderAddress,
  ...otherProps
}) => {
  const { accountAddress } = useAccountAddress();

  const tokenApprovalProps = useTokenApproval({
    token,
    spenderAddress,
    accountAddress,
  });

  return (
    <TransactionFormUi
      token={token}
      tokenNeedsToBeApproved={tokenNeedsToBeApproved}
      {...tokenApprovalProps}
      {...otherProps}
    />
  );
};

export default TransactionForm;
