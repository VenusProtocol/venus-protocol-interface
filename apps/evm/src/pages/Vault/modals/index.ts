export { default as StakeModal } from './StakeModal';
export { default as WithdrawFromVaiVaultModal } from './WithdrawFromVaiVaultModal';
export { default as WithdrawFromVestingVaultModal } from './WithdrawFromVestingVaultModal';
