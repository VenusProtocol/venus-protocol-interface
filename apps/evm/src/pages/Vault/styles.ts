import { css } from '@emotion/react';
import { useTheme } from '@mui/material';

export const useStyles = () => {
  const theme = useTheme();

  return {
    container: css`
      display: grid;
      grid-template-columns: 1fr 1fr;
      gap: ${theme.spacing(8)};

      ${theme.breakpoints.down('xl')} {
        grid-template-columns: 1fr;
      }
    `,
  };
};
