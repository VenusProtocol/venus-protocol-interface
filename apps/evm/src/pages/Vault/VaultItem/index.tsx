/** @jsxImportSource @emotion/react */
import Paper from '@mui/material/Paper';
import Typography from '@mui/material/Typography';
import BigNumber from 'bignumber.js';
import { useMemo, useState } from 'react';

import { useGetPrimeStatus } from 'clients/api';
import { Button, NoticeWarning, TokenIcon } from 'components';
import PrimeStatusBanner from 'containers/PrimeStatusBanner';
import useConvertMantissaToReadableTokenString from 'hooks/useConvertMantissaToReadableTokenString';
import { useIsFeatureEnabled } from 'hooks/useIsFeatureEnabled';
import { useLunaUstWarning } from 'packages/lunaUstWarning';
import { useTranslation } from 'packages/translations';
import { useAccountAddress } from 'packages/wallet';
import { Token } from 'types';
import { convertMantissaToTokens, formatPercentageToReadableValue } from 'utilities';

import { StakeModal, WithdrawFromVaiVaultModal, WithdrawFromVestingVaultModal } from '../modals';
import { useStyles } from './styles';
import TEST_IDS from './testIds';

type ActiveModal = 'stake' | 'withdraw';

export interface VaultItemUiProps {
  stakedToken: Token;
  rewardToken: Token;
  stakingAprPercentage: number;
  dailyEmissionMantissa: BigNumber;
  totalStakedMantissa: BigNumber;
  onStake: () => void;
  onWithdraw: () => void;
  closeActiveModal: () => void;
  isPaused: boolean;
  canWithdraw?: boolean;
  poolIndex?: number;
  activeModal?: ActiveModal;
  userStakedMantissa?: BigNumber;
  hasPendingWithdrawalsFromBeforeUpgrade?: boolean;
  className?: string;
}

export const VaultItemUi: React.FC<VaultItemUiProps> = ({
  stakedToken,
  rewardToken,
  userStakedMantissa,
  stakingAprPercentage,
  dailyEmissionMantissa,
  totalStakedMantissa,
  onStake,
  onWithdraw,
  isPaused,
  canWithdraw = true,
  activeModal,
  poolIndex,
  closeActiveModal,
  hasPendingWithdrawalsFromBeforeUpgrade,
  className,
}) => {
  const styles = useStyles();
  const { t } = useTranslation();

  const { accountAddress } = useAccountAddress();
  const isPrimeEnabled = useIsFeatureEnabled({ name: 'prime' });
  const { data: getPrimeStatusData } = useGetPrimeStatus(
    {
      accountAddress: accountAddress || '',
    },
    {
      enabled: !!accountAddress,
    },
  );
  const primePoolIndex = getPrimeStatusData?.xvsVaultPoolId;

  const readableUserStakedTokens = useConvertMantissaToReadableTokenString({
    token: stakedToken,
    value: userStakedMantissa || new BigNumber(0),
    addSymbol: false,
  });

  const dataListItems = useMemo(
    () => [
      {
        title: t('vaultItem.stakingApr', { stakeTokenName: stakedToken.symbol }),
        value: formatPercentageToReadableValue(stakingAprPercentage),
      },
      {
        title: t('vaultItem.dailyEmission'),
        value: (
          <>
            <TokenIcon css={styles.tokenIcon} token={rewardToken} />
            {convertMantissaToTokens({
              value: dailyEmissionMantissa,
              token: rewardToken,
              returnInReadableFormat: true,

              addSymbol: false,
            })}
          </>
        ),
      },
      {
        title: t('vaultItem.totalStaked'),
        value: (
          <>
            <TokenIcon css={styles.tokenIcon} token={stakedToken} />
            {convertMantissaToTokens({
              value: totalStakedMantissa,
              token: stakedToken,
              returnInReadableFormat: true,

              addSymbol: false,
            })}
          </>
        ),
      },
    ],
    [stakedToken, rewardToken, stakingAprPercentage, dailyEmissionMantissa, totalStakedMantissa, t],
  );

  return (
    <>
      <Paper css={styles.container} className={className}>
        <div css={styles.header}>
          <div css={styles.title}>
            <TokenIcon css={styles.tokenIcon} token={stakedToken} />

            <Typography variant="h4" css={styles.text} data-testid={TEST_IDS.symbol}>
              {stakedToken.symbol}
            </Typography>
          </div>
        </div>

        <Typography variant="small2" css={[styles.label, styles.stakingLabel]}>
          {t('vaultItem.youAreStake')}
        </Typography>

        <Typography
          variant="h1"
          css={styles.textStakeValue}
          data-testid={TEST_IDS.userStakedTokens}
        >
          <TokenIcon css={[styles.tokenIconLarge]} token={stakedToken} />

          {readableUserStakedTokens}
        </Typography>

        {isPrimeEnabled && primePoolIndex !== undefined && poolIndex === primePoolIndex && (
          <PrimeStatusBanner className="bg-background p-4 sm:mt-2" hidePromotionalTitle />
        )}

        <ul css={styles.dataRow}>
          {dataListItems.map(({ title, value }) => (
            <li key={title} css={styles.valueWrapper}>
              <Typography variant="small2" css={[styles.label, styles.textSmallMobile]}>
                {title}
              </Typography>

              <Typography
                variant="h4"
                css={[styles.textAligned, styles.textSmallMobile]}
                data-testid={TEST_IDS.dataListItem}
              >
                {value}
              </Typography>
            </li>
          ))}
        </ul>

        {(isPaused || hasPendingWithdrawalsFromBeforeUpgrade) && (
          <NoticeWarning
            description={
              isPaused
                ? t('vaultItem.pausedWarning')
                : t('vaultItem.blockingPendingWithdrawalsWarning')
            }
            className="mt-6"
          />
        )}

        <div css={styles.buttonsWrapper}>
          <Button
            onClick={onStake}
            css={styles.button}
            variant="primary"
            disabled={isPaused || hasPendingWithdrawalsFromBeforeUpgrade}
          >
            {t('vaultItem.stakeButton')}
          </Button>

          {canWithdraw && (
            <Button
              onClick={onWithdraw}
              css={styles.button}
              variant="secondary"
              disabled={isPaused}
            >
              {t('vaultItem.withdrawButton')}
            </Button>
          )}
        </div>
      </Paper>

      {activeModal === 'stake' && (
        <StakeModal
          stakedToken={stakedToken}
          rewardToken={rewardToken}
          handleClose={closeActiveModal}
          poolIndex={poolIndex}
        />
      )}

      {activeModal === 'withdraw' && poolIndex === undefined && stakedToken.symbol === 'VAI' && (
        <WithdrawFromVaiVaultModal handleClose={closeActiveModal} />
      )}

      {activeModal === 'withdraw' && poolIndex !== undefined && (
        <WithdrawFromVestingVaultModal
          handleClose={closeActiveModal}
          stakedToken={stakedToken}
          poolIndex={poolIndex}
          hasPendingWithdrawalsFromBeforeUpgrade={hasPendingWithdrawalsFromBeforeUpgrade || false}
        />
      )}
    </>
  );
};

export type VaultItemProps = Omit<
  VaultItemUiProps,
  'onStake' | 'onWithdraw' | 'closeActiveModal' | 'activeModal' | 'isWithdrawLoading'
>;

const VaultItem: React.FC<VaultItemProps> = ({
  stakedToken,
  rewardToken,
  poolIndex,
  ...vaultItemUiProps
}) => {
  const { userHasLunaOrUstCollateralEnabled, openLunaUstWarningModal } = useLunaUstWarning();

  const [activeModal, setActiveModal] = useState<ActiveModal | undefined>();
  const onStake = () => {
    // Block action is user has LUNA or UST enabled as collateral
    if (userHasLunaOrUstCollateralEnabled) {
      openLunaUstWarningModal();
      return;
    }

    setActiveModal('stake');
  };

  const onWithdraw = async () => {
    // Block action is user has LUNA or UST enabled as collateral
    if (userHasLunaOrUstCollateralEnabled) {
      openLunaUstWarningModal();
      return;
    }

    setActiveModal('withdraw');
  };

  const closeActiveModal = () => setActiveModal(undefined);

  return (
    <VaultItemUi
      onStake={onStake}
      onWithdraw={onWithdraw}
      activeModal={activeModal}
      closeActiveModal={closeActiveModal}
      stakedToken={stakedToken}
      rewardToken={rewardToken}
      poolIndex={poolIndex}
      // Hide withdraw button of non-vesting VRT vault when user doesn't have
      // any tokens staked in it
      canWithdraw={
        stakedToken.symbol !== 'VRT' ||
        typeof poolIndex === 'number' ||
        !vaultItemUiProps.userStakedMantissa ||
        vaultItemUiProps.userStakedMantissa.isGreaterThan(0)
      }
      {...vaultItemUiProps}
    />
  );
};

export default VaultItem;
