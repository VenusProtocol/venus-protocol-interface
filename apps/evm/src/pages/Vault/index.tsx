/** @jsxImportSource @emotion/react */
import { useGetVaults } from 'clients/api';
import { Spinner } from 'components';
import { useAccountAddress } from 'libs/wallet';
import { Vault } from 'types';

import VaultItem from './VaultItem';
import { useStyles } from './styles';

export interface VaultUiProps {
  vaults: Vault[];
  isInitialLoading: boolean;
}

const generateVaultKey = (vault: Vault) =>
  `vault-${vault.stakedToken.address}-${vault.rewardToken.address}-${vault.lockingPeriodMs || 0}`;

export const VaultUi: React.FC<VaultUiProps> = ({ vaults, isInitialLoading }) => {
  const styles = useStyles();

  if (isInitialLoading || vaults.length === 0) {
    return <Spinner />;
  }

  return (
    <div css={styles.container}>
      {vaults.map(vault => (
        <VaultItem {...vault} key={generateVaultKey(vault)} />
      ))}
    </div>
  );
};

const VaultPage: React.FC = () => {
  const { accountAddress } = useAccountAddress();
  const { data: vaults, isLoading: isGetVaultsLoading } = useGetVaults({
    accountAddress,
  });

  return <VaultUi vaults={vaults} isInitialLoading={isGetVaultsLoading} />;
};

export default VaultPage;
