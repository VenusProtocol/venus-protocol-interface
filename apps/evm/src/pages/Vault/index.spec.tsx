import Vi from 'vitest';

import { vaults as fakeVaults } from '__mocks__/models/vaults';
import { renderComponent } from 'testUtils/render';

import { useGetVaults } from 'clients/api';

import Vault from '.';
import TEST_IDS from './VaultItem/testIds';

describe('pages/Vault', () => {
  beforeEach(() => {
    (useGetVaults as Vi.Mock).mockImplementation(() => ({
      data: fakeVaults,
      isLoading: false,
    }));
  });

  it('renders without crashing', async () => {
    renderComponent(<Vault />);
  });

  it('renders vaults correctly', async () => {
    const { queryAllByTestId } = renderComponent(<Vault />);

    const symbolsElements = queryAllByTestId(TEST_IDS.symbol);
    const userStakedTokensElements = queryAllByTestId(TEST_IDS.userStakedTokens);
    const dataListItemElements = queryAllByTestId(TEST_IDS.dataListItem);

    symbolsElements.map(symbolsElement => expect(symbolsElement.textContent).toMatchSnapshot());
    userStakedTokensElements.map(userStakedTokensElement =>
      expect(userStakedTokensElement.textContent).toMatchSnapshot(),
    );
    dataListItemElements.map(dataListItemElement =>
      expect(dataListItemElement.textContent).toMatchSnapshot(),
    );
  });
});
