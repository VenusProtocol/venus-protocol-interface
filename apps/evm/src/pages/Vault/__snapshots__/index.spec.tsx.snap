// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pages/Vault > renders vaults correctly 1`] = `"VAI"`;

exports[`pages/Vault > renders vaults correctly 2`] = `"XVS"`;

exports[`pages/Vault > renders vaults correctly 3`] = `"0"`;

exports[`pages/Vault > renders vaults correctly 4`] = `"233.00"`;

exports[`pages/Vault > renders vaults correctly 5`] = `"> 10,000%"`;

exports[`pages/Vault > renders vaults correctly 6`] = `"144.00"`;

exports[`pages/Vault > renders vaults correctly 7`] = `"415.00"`;

exports[`pages/Vault > renders vaults correctly 8`] = `"12.92%"`;

exports[`pages/Vault > renders vaults correctly 9`] = `"144.00"`;

exports[`pages/Vault > renders vaults correctly 10`] = `"400.00M"`;
